components/brave_wallet/browser/brave_wallet_auto_pin_service.cc

```diff
@@ -0,0 +1,243 @@
+                                                  absl::nullopt));
+      }
+    }
+  }
+
+  for (const auto& t : known_tokens) {
+    mojom::BlockchainTokenPtr token = BraveWalletPinService::TokenFromPath(t);
+    if (token) {
+      AddOrExecute(std::make_unique<IntentData>(token, Operation::DELETE,
+                                                absl::nullopt));
+    }
+  }
+
+  CheckQueue();
+}
+
+void BraveWalletAutoPinService::PostPinToken(BlockchainTokenPtr token,
+                                             PostPinTokenCallback callback) {
+  queue_.push_back(
+      std::make_unique<IntentData>(token, Operation::ADD, absl::nullopt));
+  CheckQueue();
+}
+
+void BraveWalletAutoPinService::PostUnpinToken(BlockchainTokenPtr token,
+                                               PostPinTokenCallback callback) {
+  queue_.push_back(
+      std::make_unique<IntentData>(token, Operation::DELETE, absl::nullopt));
+  CheckQueue();
+}
+
+void BraveWalletAutoPinService::ValidateToken(
+    const std::unique_ptr<IntentData>& data) {
+  brave_wallet_pin_service_->Validate(
+      data->token->Clone(), data->service,
+      base::BindOnce(&BraveWalletAutoPinService::OnValidateTaskFinished,
+                     base::Unretained(this)));
+}
+
+void BraveWalletAutoPinService::PinToken(
+    const std::unique_ptr<IntentData>& data) {
+  brave_wallet_pin_service_->AddPin(
+      data->token->Clone(), data->service,
+      base::BindOnce(&BraveWalletAutoPinService::OnTaskFinished,
+                     base::Unretained(this)));
+}
+
+void BraveWalletAutoPinService::UnpinToken(
+    const std::unique_ptr<IntentData>& data) {
+  brave_wallet_pin_service_->RemovePin(
+      data->token->Clone(), data->service,
+      base::BindOnce(&BraveWalletAutoPinService::OnTaskFinished,
+                     base::Unretained(this)));
+}
+
+void BraveWalletAutoPinService::AddOrExecute(std::unique_ptr<IntentData> data) {
+  DCHECK(data);
+  for (const auto& v : queue_) {
+    if (v->token == data->token && v->service == data->service) {
+      return;
+    }
+  }
+  if (current_ && current_->token == data->token &&
+      current_->service == data->service) {
+    return;
+  }
+  if (data->operation == Operation::ADD) {
+    brave_wallet_pin_service_->MarkAsPendingForPinning(data->token,
+                                                       data->service);
+  } else if (data->operation == Operation::DELETE) {
+    brave_wallet_pin_service_->MarkAsPendingForUnpinning(data->token,
+                                                         data->service);
+  }
+  queue_.push_back(std::move(data));
+  CheckQueue();
+}
+
+void BraveWalletAutoPinService::PostRetry(std::unique_ptr<IntentData> data) {
+  int multiply = ++data->attempt;
+  base::SequencedTaskRunnerHandle::Get()->PostDelayedTask(
+      FROM_HERE,
+      base::BindOnce(&BraveWalletAutoPinService::AddOrExecute,
+                     base::Unretained(this), std::move(data)),
+      base::Minutes(1 * multiply));
```

Comment:
It is unsafe to pass Unretained(this) to PostDelayedTask. Should pass weak ptr instead